### PR TITLE
BUDGET APP

### PR DESCRIPTION
**Pull request by NATASHA-ct**
#### Design
- You should follow these [design guidelines](https://www.behance.net/gallery/19759151/Snapscan-iOs-design-and-branding?tracking_source=), including:
  - Colors.
  - Typography: font face, size and weight.
  - Layout: composition and space between elements.

> NOTE: In these design guidelines there are several UIs that you won't need for this exercise; also, some pages are not given a design and you will create them following the design guidelines of the other pages.
Original design idea by [Gregoire Vella on Behance](https://www.behance.net/gregoirevella).

The [Creative Commons license of the design](https://creativecommons.org/licenses/by-nc/4.0/) requires that you give appropriate credit to the author. Therefore, you must do it in the README of your project.

#### Interactions
- Splash screen
  - A simple page with the name of your app (yes, you need to choose one), and links to the sign up and log in pages.

- Sign up and log in pages
  - The user should be able to register in the app with full name, email and password (all mandatory).
  - The user can log into the app using email and password.
  - If the user is not logged in, they can't access pages that require the user to be logged in (all the pages described below).

- Home page (catagories page)
  - When the user logs in, they are presented with the categories page.
  - For each category, the user can see their name, icon and the total amount of all the transactions that belongs to that category.
  - When the user clicks (or taps) on a category item, the application navigates to the transactions page for that category.
  - There is a button "add a new category" at the bottom that brings the user to the page to create a new category.

- Transactions page
  - For a given category, the list of transactions is presented, ordered by the most recent.
  - At the top of the page the user could see the total amount for the category (sum of all of the amounts of the transactions in that category).
  - There is a button "add a new transaction" at the bottom that brings the user to the page to create a new transaction.
  - When the user clicks on the "Back" button (<), the user navigates to the home page.

- "Add a new category" page
  - The user fills out a form to create a new category, indicating their name and icon (both mandatory).
  - The user click (or taps) the "Save" button to create the new category, and is taken to the home page on success.
  - When the user clicks on the "Back" button (<), the user navigates to the home page.

- "Add a new transaction" page
  - The user fills out a form to create a new transaction with:
    - name (mandatory)
    - amount (mandatory)
    - categories (mandatory at least one)
  - The user click (or taps) the "Save" button to create the new transation, and is taken to the transactions page for that category.
  - When the user clicks on the "Back" button (<), the user navigates to the transactions page for that category.

#### Testing requirements
- Create unit and integration tests for all the most important components of your RoR application.

#### Technical requirements

- You should use Postgres as your database.
- You should use devise for authentication.
- You should validate all user input to make sure that anyone with bad intentions cannot compromise your app.
- You can use a view template engine of your choice (.erb, .slim, .haml).
- The project should be deployed and accessible online.
- Your database schema should reflect the following structure:

**LOOM VIDEO**
https://www.loom.com/share/32eb82e0582146cd8ab49ccd573ed667

### CODE REVIEW

File: app/controllers/expenses_controller.rb [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+  # GET /expenses/1/edit
+  def edit
+    @group = current_user.groups.find_by(id: params[:group_id])
+    @expense = @group.expenses.find(params[:id])
+  end
+
+  # POST /expenses or /expenses.json
+  def create
+    @group = current_user.groups.find_by(id: params[:group_id])
```
Comment (AshakaE) [inline app/controllers/expenses_controller.rb]: The use of ` @group = current_user.groups.find(params[:group_id])` across your controller violates the DRY principle. You can declare it once at instance level `def new` and use everywhere in you controller

- [x] Kindly refactor your controller

**Review by AshakaE — CHANGES_REQUESTED** [review]
Hi @NATASHA-ct 👋  , 

Good job so far
There are some issues that you still need to work on to go to the next project but you are almost there!
<FEEL FREE TO CHANGE THIS MESSAGE TO YOUR PERSONAL WORDS>

## Highlights
- Presentation was great ✨ 👏 👏 
- Git flow ✅ 
- README ✅ 


## Required Changes ♻️
_Check the comments under the review._

- [x] Kindly deploy your project and update your README with the live demo link
- In your presentation, you missed out the part where you'd highlight about a piece of the code you found interesting
- [x] Kindly make changes to your video presentation 🙏 

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/controllers/expenses_controller.rb (hunk 1 above)
Comment (NATASHA-ct) [reply app/controllers/expenses_controller.rb]: - [ ] Hello @AshakaE thank you for your review, kindly explain further what exactly I should fix on the video.

- [ ] also for Heroku, it is no longer offering free membership and my peers didn't manage to deploy as well, if you are available to assist can we have a zoom call?

Comment (AshakaE) [reply app/controllers/expenses_controller.rb]: With respect to your video presentation, what I meant is `The video contains a highlight about a piece of the code the student finds interesting or is proud of making.`

Comment (NATASHA-ct) [reply app/controllers/expenses_controller.rb]: Hello @AshakaE  thank you for pointing out refactoring but this is not duplication , I tried to make the refactoring but it gives me an error .
image

File: app/controllers/expenses_controller.rb [hunk 2]
```diff
@@ -0,0 +1,81 @@
+class ExpensesController < ApplicationController
+  before_action :authenticate_user!
+  before_action :set_expense, only: %i[show edit update destroy]
+
+  # GET /expenses or /expenses.json
+  def index
+    @group = current_user.groups.find(params[:group_id])
```
Comment (SajjadAhmad14) [inline app/controllers/expenses_controller.rb]: Please refactor this piece of code when you find a group by group id which is used in almost every action of the controller. One way is to define a private method  and use `before_action` filter as you are using for `#set_expense`

File: README.md [hunk 3]
```diff
@@ -1,285 +1,125 @@
...
 
-<div align="center">
+- [Log in and registration of the user]
+- [Recipe list ,Food list, Public recipe list pages]
+- [Functionality to Generate a shopping list with total cost]
 
-  <img src="murple_logo.png" alt="logo" width="140"  height="auto" />
-  <br/>
+## Live Demo
+- To be deployed
```
Comment (SajjadAhmad14) [inline README.md]: Please deploy the app online as it's requested by previous code reviewer. You can use any other service to host it and one of them is Netlify.

**Review by SajjadAhmad14 — CHANGES_REQUESTED** [review]
## Required Changes ♻️
> "Change is inevitable. Change is constant. -Benjamin Disraeli"
## Hi Tascha,
### Great work on making the changes requested by a previous reviewer 👏 
### You've done well implementing some of the requested changes, but there are still some which aren't been addressed yet.
## To Highlight 🎉
PR has a titel and summary. :heavy_check_mark:
UI is friendly and facianating. :heavy_check_mark:
## To Improve 🙏

- [ ] Please check inline comments under the review

## Optional suggestions
_Comments with the [OPTIONAL] prefix are not a must to implement however, It's strongly recommended to work on them as they can make your code better._

## Happy coding!👏👏👏

#### Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
#### Please remember to tag me in your question so I can receive the notification.
**_Note: Please do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s).

File: app/controllers/expenses_controller.rb (hunk 2 above)
Comment (NATASHA-ct) [reply app/controllers/expenses_controller.rb]: Thank you @SajjadAhmad14  for your review, I also mentioned to the previous reviewer that when I try this it's giving me errors, on my expense page. Please kindly assist if you are available

image

File: README.md (hunk 3 above)
Comment (NATASHA-ct) [reply README.md]: Thank you for this, but Netlify is not deploying ROR projects
image

File: app/controllers/expenses_controller.rb [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    @expenses = @group.expenses..order(created_at: :desc)
+  end
+
+  # GET /expenses/1 or /expenses/1.json
+  def show; end
+
+  # GET /expenses/new
+  def new
+    @group = current_user.groups.find_by(id: params[:group_id])
```
Comment (EricMbouwe) [inline app/controllers/expenses_controller.rb]: - [x] As mentioned above please refactor the `groups.find_by(id: params[:group_id])` piece of code when you find a group by group id which is used in almost every action of the controller. This makes your code much cleaner and robust

**Review by EricMbouwe — CHANGES_REQUESTED** [review]
Hi @NATASHA-ct, 

Great work on making the changes 👏⚡
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.


# Suggested changes ♻
_Check the comments under the review._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.


**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

**Review by elisha2kyakpo1 — APPROVED** [review]
Hi 👋 @NATASHA-ct, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Highlights

✔️ You have done a great job implementing the changes requested by the previous reviewer👏

## Optional suggestions

- [x] The project UI still requires to be worked on to meet the project requirements.
- [x] Since heroku will be unavailable starting 28th of November 2022, you should consider using a different infrastructure platform.

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**